### ISSUE TITLE
Feature for sorting

### ISSUE THREAD

**Issue by Aryaarrya** [facebookresearch/CrypTen#293]
##  Feature

I was wondering how we can sort a crypten tensor since we can't apply conditional expressions on CrypTensors. 

Something like

```python
x_enc = crypten.cryptensor([3.0, 1.0, 2.0])
y_enc = x_enc.sort()
```

**Comment by knottb** [comment]
We don't currently have an implementation for sorting algorithms. This could be implemented using multiplexed swapping of elements, but would take some work to optimize.

e.g.

```
def swap_ascending(x_enc, ind1, ind2):
    assert ind1 < ind2  # Assuming ascending from ind1 to ind2

    is_smaller = x_enc[ind1] < x_enc[ind2]
    smaller_value = is_smaller * x_enc[ind1] + (1 - is_smaller) * x_enc[ind2]
    larger_value = x_enc[ind1] + x_enc[ind2] - smaller_value

    x_enc[ind1] = smaller_value
    x_enc[ind2] = larger_value
```

This can then be composed to implement many sorting algorithms.

**Comment by lvdmaaten** [comment]
Given that this is constantly converting between arithmetic and binary secret sharing, this would be very slow, right? It seems to me that there must be ways to reduce the number of conversions substantially by doing more comparisons for each conversion?

**Comment by knottb** [comment]
Yes, this is likely to be fairly slow as there are a lot of sequential comparisons. It is possible that many could be done in parallel, but I'd have to think about whether there is a sorting algorithm that vectorizes well.
